File: src/rwe/util.test.cpp
#include <catch2/catch_test_macros.hpp>
#include <rwe/util.h>
#include <set>

namespace rwe
{
    TEST_CASE("the user data directory follows the XDG base directory rule", "[util]")
    {
        namespace fs = std::filesystem;

        SECTION("without XDG_DATA_HOME the data home is ~/.local/share, and ~/.rwe is the fallback")
        {
            auto c = localDataPathCandidates(nullptr, "/home/someone");
            REQUIRE(c == std::vector<fs::path>{"/home/someone/.local/share/rwe", "/home/someone/.rwe"});
        }

        SECTION("an absolute XDG_DATA_HOME is honoured")
        {
            auto c = localDataPathCandidates("/mnt/data", "/home/someone");
            REQUIRE(c.front() == fs::path("/mnt/data/rwe"));
            REQUIRE(c.back() == fs::path("/home/someone/.rwe"));
        }

        SECTION("an empty or relative XDG_DATA_HOME is invalid, and the default applies")
        {
            REQUIRE(localDataPathCandidates("", "/home/someone").front() == fs::path("/home/someone/.local/share/rwe"));
            REQUIRE(localDataPathCandidates("share", "/home/someone").front() == fs::path("/home/someone/.local/share/rwe"));
        }

        SECTION("no HOME means no data directory at all")
        {
            REQUIRE(localDataPathCandidates(nullptr, nullptr).empty());
            REQUIRE(localDataPathCandidates("/mnt/data", "").empty());
            REQUIRE_FALSE(chooseLocalDataPath({}, [](const fs::path&) { return true; }).has_value());
        }

        SECTION("an existing ~/.rwe keeps being used until the new location exists")
        {
            auto c = localDataPathCandidates(nullptr, "/home/someone");
            auto existing = [](std::set<fs::path> dirs) {
                return [dirs = std::move(dirs)](const fs::path& p) { return dirs.count(p) != 0; };
            };

            // A fresh machine: the new place.
            REQUIRE(*chooseLocalDataPath(c, existing({})) == fs::path("/home/someone/.local/share/rwe"));
            // An old install: where its files are.
            REQUIRE(*chooseLocalDataPath(c, existing({"/home/someone/.rwe"})) == fs::path("/home/someone/.rwe"));
            // Both, after someone has moved: the new place wins.
            REQUIRE(*chooseLocalDataPath(c, existing({"/home/someone/.rwe", "/home/someone/.local/share/rwe"})) == fs::path("/home/someone/.local/share/rwe"));
        }
    }
}

File: launcher/src/launcher/util.test.ts
import path from "path";
import { chooseRweUserPath, rweUserPathCandidates } from "./util";

describe("rweUserPathCandidates", () => {
  it("uses %APPDATA%\\RWE on Windows", () => {
    expect(
      rweUserPathCandidates("win32", { APPDATA: "C:\\Users\\x\\AppData\\Roaming" })
    ).toEqual([path.join("C:\\Users\\x\\AppData\\Roaming", "RWE")]);
  });
  it("throws without APPDATA on Windows", () => {
    expect(() => rweUserPathCandidates("win32", {})).toThrow();
  });
  it("defaults to ~/.local/share/rwe with ~/.rwe as the fallback", () => {
    expect(rweUserPathCandidates("linux", { HOME: "/home/someone" })).toEqual([
      "/home/someone/.local/share/rwe",
      "/home/someone/.rwe",
    ]);
  });
  it("honours an absolute XDG_DATA_HOME and ignores an invalid one", () => {
    expect(
      rweUserPathCandidates("linux", {
        HOME: "/home/someone",
        XDG_DATA_HOME: "/mnt/data",
      })[0]
    ).toEqual("/mnt/data/rwe");
    expect(
      rweUserPathCandidates("linux", {
        HOME: "/home/someone",
        XDG_DATA_HOME: "share",
      })[0]
    ).toEqual("/home/someone/.local/share/rwe");
    expect(
      rweUserPathCandidates("linux", { HOME: "/home/someone", XDG_DATA_HOME: "" })[0]
    ).toEqual("/home/someone/.local/share/rwe");
  });
  it("throws without a home directory", () => {
    expect(() => rweUserPathCandidates("linux", {})).toThrow();
  });
});

describe("chooseRweUserPath", () => {
  const candidates = ["/home/someone/.local/share/rwe", "/home/someone/.rwe"];
  it("picks the new place on a fresh machine", () => {
    expect(chooseRweUserPath(candidates, () => false)).toEqual(candidates[0]);
  });
  it("keeps an existing ~/.rwe until the new place exists", () => {
    expect(
      chooseRweUserPath(candidates, (p) => p === "/home/someone/.rwe")
    ).toEqual("/home/someone/.rwe");
  });
  it("prefers the new place once it exists", () => {
    expect(chooseRweUserPath(candidates, () => true)).toEqual(candidates[0]);
  });
});
